dbt_subprojects/daily_spellbook/macros/models/sector/addresses/addresses_info.sql

```diff
@@ -0,0 +1,94 @@
+{% macro addresses_info(blockchain, transactions, token_transfers, creation_traces, first_funded_by, contracts) %}
+
+{% if not is_incremental() %}
+
+WITH executed_txs AS (
+    SELECT "from" AS address
+    , COUNT(*) AS executed_tx_count
+    , COALESCE(MAX(nonce), 0) AS max_nonce
+    , MAX(block_time) AS last_tx_block_time
+    , MAX(block_number) AS last_tx_block_number
+    FROM {{transactions}}
+    GROUP BY 1
+    )
+
+, is_contract AS (
+    SELECT ct.address
+    , true AS is_smart_contract
+    , c.namespace
+    , c.name
+    FROM {{creation_traces}} ct
+    LEFT JOIN {{contracts}} c ON ct.address=c.address
+    )
+
+SELECT '{{blockchain}}' AS blockchain
+, address
+, executed_tx_count
+, max_nonce
+, COALESCE(is_smart_contract, false) AS is_smart_contract
+, namespace
+, name
+, last_tx_block_time AS last_seen
+, last_tx_block_number
+FROM executed_txs
+LEFT JOIN is_contract USING (address)
+
+
+
+{% else %}
+
+
+
+WITH executed_txs AS (
+    SELECT txs."from" AS address
+    , COUNT(*) AS executed_tx_count
+    , MAX(txs.nonce) AS max_nonce
+    , MAX(txs.block_time) AS last_tx_block_time
+    , MAX(txs.block_number) AS last_tx_block_number
+    FROM {{transactions}} txs
+    LEFT JOIN {{this}} t ON txs."from"=t.address
+        AND txs.block_number>t.last_tx_block_number
+    WHERE {{ incremental_predicate('txs.block_time') }}
+    GROUP BY 1
+    )
+
+, is_contract AS (
+    SELECT ct.address
+    , true AS is_smart_contract
+    , c.namespace
+    , c.name
+    FROM {{creation_traces}} ct
+    LEFT JOIN {{this}} t ON ct.address=t.address
+        AND ct.block_number>t.last_tx_block_number
+    LEFT JOIN {{contracts}} c ON ct.address=c.address
+    WHERE {{ incremental_predicate('ct.block_time') }}
+    )
+
+, new_data AS (
+    SELECT address
+    , executed_tx_count
+    , max_nonce
+    , is_smart_contract
+    , namespace
+    , name
+    , last_tx_block_time
+    , last_tx_block_number
+    FROM executed_txs
+    LEFT JOIN is_contract USING (address)
+    )
+
+SELECT '{{blockchain}}' AS blockchain
+, nd.address
+, nd.executed_tx_count + t.executed_tx_count AS executed_tx_count
+, COALESCE(nd.max_nonce, t.max_nonce) AS max_nonce
+, COALESCE(COALESCE(nd.is_smart_contract, t.is_smart_contract), false) AS is_smart_contract
+, COALESCE(nd.namespace, t.namespace) AS namespace
+, COALESCE(nd.name, t.name) AS name
+, GREATEST(nd.last_tx_block_time, t.last_seen) AS last_seen
+, GREATEST(nd.last_tx_block_number, t.last_tx_block_number) AS last_tx_block_number
+FROM new_data nd
+LEFT JOIN {{this}} t ON t.address=nd.address
```

Comment:
(commenting on this line because i believe this join to `{{ this }}` is likely all that's needed)

i think we're close, but not quite accomplishing what you're looking for. i still think the left joins in CTEs above to the target table aren't doing anything, even with the block number filter. the `count(*) as executed_tx_count` is still just performing on the base transactions table filter to the last one day. once you add that value to existing count in line 82, you will likely be double counting tx's each day.

maybe the move is to introduce a CTE which reads from `{{ this }}` and grabs max block number as the only output. you could then potentially inner join to this new CTE in `executed_txs` and filter where source block number is greater than the target in the conditions. does that make sense?

we can probably remove the left join in `is_contract` altogether. the merge statement behind the scenes should handle these updates for us. i believe the only field requiring lookup to target table is the count

Reply:
I'm not sure I fully understand, the `new_data` cte is fetched from the `executed_txs` cte which has:

```
LEFT JOIN {{this}} t ON txs."from"=t.address
    AND txs.block_number>t.last_tx_block_number
```

afaik, that should mean that it only adds transactions which are past the `last_tx_block_number` value that gets updated every time there's new data. where would you see the double counting?

Reply:
let me share a query that hopefully visualizes what i'm saying. if you run the below, you will notice that the current incremental setup will return full tx_count for the entire one last day (our incremental predicate outputs column passed in minus one day). the left join to target table with a filter on block number helps to null out values from the target, however, the base table which performs the left join will retain all of it's data. that means all 300k+ transactions in last one day. if you run this model multiple times in a row, it's going to add that value over and over (or something close to 300k, as timestamps will slightly shift each run).

```
WITH
  executed_txs AS (
    SELECT
      "from" AS address,
      COUNT(*) AS executed_tx_count,
      COALESCE(MAX(nonce), 0) AS max_nonce,
      MIN(block_time) AS first_tx_block_time,
      MAX(block_time) AS last_tx_block_time,
      MIN(block_number) AS first_tx_block_number,
      MAX(block_number) AS last_tx_block_number
    FROM
      "delta_prod"."arbitrum"."transactions"
    where
      "from" = 0x00000000000000000000000000000000000a4b05 --example address
      and block_time <= timestamp '2024-09-16 17:14' --block time at runtime of historical data
      and block_time >= timestamp '2024-09-15 17:14' --incremental filter reads from last 1 day
    GROUP BY
      1
  )
select
  txs.*,
  t.address as target_lookup_value --this is null due to left join with block number filter
from
  executed_txs as txs
  LEFT JOIN test_schema.git_dunesql_99fe47c_addresses_arbitrum_info t ON txs.address = t.address
  AND txs.last_tx_block_number > t.last_tx_block_number
```